File: Behavioral/Lab23/C#/Entities/TaxCalculatorVisitor.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Interfaces;

namespace C_.Entities
{
    public class TaxCalculatorVisitor : IEstateVisitor
    {
        public void Visit(RealEstate estate){
            switch (estate){
                case House house:
                    VisitHouse(house);
                    break;
                case Apartment apartment:
                    VisitApartment(apartment);
                    break;
                case Room room:
                    VisitRoom(room);
                    break;
                default:
                    throw new Exception("Unknown RealEstate type");
            }
        }
        protected void VisitHouse(House house)
        {
            double tax = house.Area * 1.5; 
            Console.WriteLine($"House at {house.Address}: Tax is {tax}");
        }

        protected void VisitApartment(Apartment apartment)
        {
            double tax = apartment.Area * 1.2; 
            Console.WriteLine($"Apartment at {apartment.Address}: Tax is {tax}");
        }

        protected void VisitRoom(Room room)
        {
            double tax = room.Area * 0.8;
            Console.WriteLine($"Room at {room.Address}: Tax is {tax}");
        }
    }

}

File: Behavioral/Lab23/Typescript/src/Entities/tax_calculator_visitor.ts
import { IEstateVisitor } from "../Interfaces/estate_visitor";
import { Apartment } from "./apartment";
import { House } from "./house";
import { RealEstate } from "./real_estate";
import { Room } from "./room";


export class TaxCalculatorVisitor implements IEstateVisitor {
    
    public visit(estate: RealEstate): void {
        const className = estate instanceof House ? "House" : estate instanceof Apartment ? "Apartment" : "Room";
        switch (className) {
            case ("House"):
                this.visitHouse(estate as House);
                break;
            case ("Apartment"):
                this.visitApartment(estate as Apartment);
                break;
            case ("Room"):
                this.visitRoom(estate as Room);
                break;
            default:
                throw new Error("Unknown RealEstate type");
        }
    }

    protected visitHouse(house: House): void {
        const tax = house.area * 1.5;
        console.log(`House at ${house.address}: Tax is ${tax.toLocaleString('en-US', {style: 'currency', currency: 'USD'})}`);
    }

    protected visitApartment(apartment: Apartment): void {
        const tax = apartment.area * 1.2;
        console.log(`Apartment at ${apartment.address}: Tax is ${tax.toLocaleString('en-US', {style: 'currency', currency: 'USD'})}`);
    }

    protected visitRoom(room: Room): void {
        const tax = room.area * 0.8;
        console.log(`Room at ${room.address}: Tax is ${tax.toLocaleString('en-US', {style: 'currency', currency: 'USD'})}`);
    } 
}